text/0000-nested-publish.md

```diff
@@ -0,0 +1,183 @@
+- Feature Name: `nested_publish`
+- Start Date: 2023-06-30
+- RFC PR: [rust-lang/rfcs#3452](https://github.com/rust-lang/rfcs/pull/3452)
+- Rust Issue: ...
+
+# Summary
+[summary]: #summary
+
+Allow Cargo packages to be bundled within other Cargo packages when they are published (not just in unpublished workspaces).
+
+# Motivation
+[motivation]: #motivation
+
+There are a number of reasons why a Rust developer currently may feel the need to create multiple library crates, and therefore multiple Cargo packages (since one package contains at most one library crate). These multiple libraries could be:
+
+* A trait declaration and a corresponding derive macro (which must be defined in a separate proc-macro library).
+* A library that uses a build script that uses another library or binary (e.g. for precomputation or bindings generation).
+* A logically singular library broken into multiple parts to speed up compilation.
+
+Currently, developers must publish these packages separately. This has several disadvantages (see the [Rationale](#rationale-and-alternatives) section for further details):
+
+* Clutters the public view of the registry with packages not intended to be usable on their own, and which may even become obsolete as internal architecture changes.
+* Requires multiple `cargo publish` operations (this could be fixed with bulk publication) and writing public metadata for each package.
+* Can result in semver violations and thus compilation failures, due to the developer not thinking about semver compatibility within the group.
+
+This RFC will allow developers to avoid all of these inconveniences and hazards by publishing a single package.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+By default (and always, prior to this RFC's implementation):
+
+* If your package contains any sub-packages, Cargo [excludes](https://doc.rust-lang.org/cargo/reference/manifest.html#the-exclude-and-include-fields) them from the `.crate` archive file produced by `cargo package` and `cargo publish`.
+* If your package contains any non-`dev` dependencies which do not give a `version = "..."`, it cannot be published to `crates.io`.
```

Comment:
We should also call out the dev-dependcency behavior of stripping them if they lack `version` fields.

Reply:
I wasn't aware that was exactly what happened, but I can see how it must be.